### ISSUE TITLE
Choose consistent time format

### ISSUE THREAD

**Issue by reece** [ga4gh/schemas#568]
#552 implemented ISO8601 for the variation api. CallSet, Variant, and ReadGroup have "created" and "updated" fields, but these were unchanged.

This proposal is switch remaining timestamps to ISO8601 and rename fields to createdDateTime and updatedDateTime for consistency.

@jeromekelleher, @david4096, @diekhans, @andrewjesaitis : Comments appreciated since you three probably know which monsters this will tickle in our code.

**Comment by mbaudis** [comment]
+1

@reece Probably no monsters to be woken, since these attributes haven't been really used.

**Comment by jeromekelleher** [comment]
Here are my thoughts:

1. We should use iso8601 everywhere for these timestamps;
2. I don't think this will cause any implementation hassles (other than actually doing it);
3. I'm not convinced by the need for the name change to `createdDateTime`. What's wrong with `created` and `updated`? We don't feel the need to put the types into the names for other fields (startLong, nameString), so why put them in here? I realise I'm probably joining the party late on this one though, so sorry if we're going over old ground on this argument!

@calbach, what's your thoughts on this, are you OK with the shift to ISO8601?

**Comment by reece** [comment]
I'm happy with any of the likely outcomes that promotes consistency in type and naming. 

My weak preference is ISO8601 and updated/created.

(I had suggested DateTime previously, but only as a reaction to updateTIme, which seemed incorrectly suffix typed to me.)

**Comment by mbaudis** [comment]
@jeromekelleher While I'm fine with ```created``` | ```updated```, this was seen as ambiguous in var. discussions ("updated" could be a Boolean/flag) & so @diekhans switched it to the current.

I had started the whole ISO8601 switch & [documentation](http://ga4ghdata.org/Sites/github/mbaudis/biosample/html/api/metadata.html#date-and-time-format-specifications), but very explicitly called those ```recordCreateTime``` | ```recordUpdateTime``` ... But as said, all fine with me, since the documentation should make it clear.

**Comment by diekhans** [comment]
My thinking on *DateTime isn't they type (which was Unix time,
now ISO time), but clarity on the meaning of the field.
`created' sounds like a boolean to me, but what ever, as long as
it is consistent.

**Comment by jeromekelleher** [comment]
I think in the interest of keeping things simple we should maintain 'created' and 'updated', as changing the names will cause friction for existing implementations. I think the advantages of `createdDateTime` over `created` are pretty subjective.

**Comment by diekhans** [comment]
I agree it's somewhat subjective, although we are changing the
value, so they will suffer anyway we do it.  The important thing
is to come up something that makes sense in the long term.

Probably a more important thing is to look at the meaning.  If
someone am offering up date from another archive (say NCBI),
what does created time mean anyway?  Is it the time imported or
the time created at NCBI.  Both have use cases and they are not
interchangeable.

Michael had recordCreated at one time; I wasn't crazy about
this, because we had nothing to contrast it against.  However,
maybe before we decide on names, we should rigorously define
meaning.

**Comment by jeromekelleher** [comment]
If I'm honest @diekhans, I think we'd be better off just deleting all these timestamp fields until there is an actual documented problem that they are supposed to solve. To me, this was another "we'll probably need something like this at some point, so let's chuck it in" part of the API. The reference server doesn't do anything of any use with them, and I can't really see that changing in the near future.

Does anybody share this view? Not existing at all is a very well documented and perfectly standardised across all objects!

**Comment by mbaudis** [comment]
@jeromekelleher While you have a seemingly valid & self contained argument there, this is in the same direction that has led to the non-implementation of dearly needed objects & attributes. I, personally, don't care for reads; variants/calls + biodata / metadata are fine with me (and the vast majority of potential users).
So, lets not get hung up on names & *keep* the stuff in that certainly will be needed, is well defined and is not expected to clash with any further developments. I'm getting dizzy from moving in circles.

**Comment by jeromekelleher** [comment]
Fair enough @mbaudis, just testing the water. I'm happy to accept that these timestamps are potentially useful.  So, let's keep the names, and move to ISO?

**Comment by mbaudis** [comment]
@jeromekelleher Thanks & yes! And the person who first gets her/his version of consistent naming accepted in a schema wide PR wins - as long as it is ISO ;-)

**Comment by diekhans** [comment]
Note, we have fields like:  ReadStats.alignedReadCount
while is analogous to createDateTime.

But my big concern is this:

ReadGroup.created is documented as

   "The time at which this read group was created in
    milliseconds from the epoch."

Does this mean when the readgroup was loaded into the server or
when the read group was sequenced?  This is entirely too underspecified
to be useful.

Use cases:

   - Find out what has been added to a sever in the last week
vs
   - Find readgroups that are more than 5 years old.

**Comment by diekhans** [comment]
@calbach what does create/update time mean in the Google API?  Is it the time the data is
loaded into the server or the time the data was actually created?

**Comment by calbach** [comment]
@diekhans The time it was loaded into the server.

@jeromekelleher thanks, sorry I missed the summon..

I've seen a few PRs going by, but is there an explanation somewhere of why this change is being made? I see https://github.com/ga4gh/schemas/pull/552, but I don't see any rationale provided for the change or any issues linked which might contain this rationale. This clearly complicates the spec, so some justification should be provided.

Note that proto3 has a [Timestamp type](https://developers.google.com/protocol-buffers/docs/proto3#json) which will be emitted as RFC 3339 when converted to JSON. Might simplify client/server implementers' lives to just use that.

**Comment by diekhans** [comment]
Thanks @calbach

We really need more specification off all timestamp.  Loaded in
server is not always what is needed when using a pre-existing
dataset.

I guess the question is: what are the use cases for timestamps
in GA4GH?

The rationale is to use an single international standard for all
times in GA4GH.  Having multiple time formats is confusing.
POSIX time is not suitable for may types of data/time
representation.

As far as I can tell, RFC 3339 is ISO 8601, so protobuf support
is a big plus.

**Comment by mbaudis** [comment]
@diekhans +1

@calbach:
There are some areas for which standard formats have to be defined, before anybody even thinks about an implementation. And bad if format clashes exist in some corner of schema space. Probabaly one could define the reference DNA positions in imperial units.
Clear cases for defining formats/names are:

* object naming conventions
* names of omnipresent attributes (e.g. "id" type of attributes)
* character use/encoding (e.g. URI encoding, word characters for referenced object identifiers)
* international standard formats (e.g. here the ISO time vs. the "imperial" epoch)
...

I have put some work into such attempts, e.g. in [starting "id" discussions](https://github.com/ga4gh/schemas/issues/418) and the ISO8601 writeup. While there can be discussions about how to name a time attribute, the format is a closed case. Obviously, it is absolutely open how things are done locally; however, any time value exposed in a GA4GH compatible schema has to be ISO8601. That is also the scope of the @diekhans question about

     what does create/update time mean in the Google API?

... which tries to avoid clashes in the format, not the name (Does Google really use epoch? Should fix it soon ...).

**Comment by calbach** [comment]
> the format is a closed case

My question was for documentation of or rationale behind this decision. I still don't see context provided anywhere in this issue or in any of the linked issues/PRs. I'm not necessarily disagreeing with the proposed change - I'm asking for it to be clearly explained/documented.

> We really need more specification off all timestamp.

Agreed. In general I would expect a generic field like `created` or `updated` to refer to the resource as it exists in the server, though this should be made clear in the documentation if it isn't already.

**Comment by diekhans** [comment]
The documentation on time format is here:

    https://github.com/ga4gh/schemas/blob/master/doc/source/api/metadata.rst

but it's just a summary of ISO8601.  Since this is a issue, not
a PR, it would be a great place to collect rationale.  The
rationale should end up in the documentation.

From 15 years of dealing with bioinformatics data, would assume
the created data is when the data was original made available.
So if it was genbank data made available though a GA4GH server,
I would expect it to be the Genbank date.  I would really rather
not have generically named fields that can be interpreted in
multiple ways.

**Comment by david4096** [comment]
Will close with https://github.com/ga4gh/schemas/pull/690
